File: 26-remove-duplicates-from-sorted-array/26-remove-duplicates-from-sorted-array.java
/*
 * @lc  id=26 lang=java
 *
 * [26] Remove Duplicates from Sorted Array
 */

 /**
  * 
  Given a sorted array nums, remove the duplicates in-place such that each element appears only once and returns the new length.

Do not allocate extra space for another array, you must do this by modifying the input array in-place with O(1) extra memory.

Example 1:

Given nums = [1,1,2],

Your function should return length = 2, with the first two elements of nums being 1 and 2 respectively.

It doesn't matter what you leave beyond the returned length.
Example 2:

Given nums = [0,0,1,1,1,2,2,3,3,4],

Your function should return length = 5, with the first five elements of nums being modified to 0, 1, 2, 3, and 4 respectively.

It doesn't matter what values are set beyond the returned length.
  */

/**
 * ## Solution intuitive:
 * - find the first unmatch, move following
 * - if cur equal to previous, move following
 * 
 * ## a big problem is too many unnecessary move
 * - need to maintain counter, idx which are error-prone
 */
class Solution {
    public int removeDuplicates(int[] nums) {
        int len = nums.length;
        int offset = 0;
        int idx = 0; 
        while(idx+offset < len -1) {
            if(nums[idx] == nums[idx+1]) {
                int left = idx;
                int right = idx + 1;
                while(right<len) {
                    nums[left++] = nums[right++];
                }
                offset++;
            } else {
                idx++;
            }
        }
        return len - offset ;
    }
}
// @lc code=end




/*
 * @lc id=26 lang=java
 *
 * [26] Remove Duplicates from Sorted Array
 */
/**
 * Solution clever:
 * - two pointers
 * - one for pivot, one for scanning
 */
// @lc code=start
class Solution {
    public int removeDuplicates(int[] nums) {
        int len = nums.length;
        if(len < 2) {
            return len;
        }

        int idx = 1;
        for(int i = 1; i<len; i++) {
            if(nums[i] != nums[i-1]) {
                nums[idx++] = nums[i];
            }
        }
        return idx;

    }
}
// @lc code=end

File: algorithm/26.remove-duplicates-from-sorted-array/26.remove-duplicates-from-sorted-array.py
#
# @lc app=leetcode id=26 lang=python3
#
# [26] Remove Duplicates from Sorted Array
#

"""
Remove Duplicates from Sorted Array
Category	Difficulty	Likes	Dislikes
algorithms	Easy (59.33%)	16177	19537
Tags
Companies
Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique element appears only once. The relative order of the elements should be kept the same. Then return the number of unique elements in nums.

Consider the number of unique elements of nums to be k, to get accepted, you need to do the following things:

Change the array nums such that the first k elements of nums contain the unique elements in the order they were present in nums initially. The remaining elements of nums are not important as well as the size of nums.
Return k.
Custom Judge:

The judge will test your solution with the following code:

int[] nums = [...]; // Input array
int[] expectedNums = [...]; // The expected answer with correct length

int k = removeDuplicates(nums); // Calls your implementation

assert k == expectedNums.length;
for (int i = 0; i < k; i++) {
    assert nums[i] == expectedNums[i];
}
If all assertions pass, then your solution will be accepted.

 

Example 1:

Input: nums = [1,1,2]
Output: 2, nums = [1,2,_]
Explanation: Your function should return k = 2, with the first two elements of nums being 1 and 2 respectively.
It does not matter what you leave beyond the returned k (hence they are underscores).
Example 2:

Input: nums = [0,0,1,1,1,2,2,3,3,4]
Output: 5, nums = [0,1,2,3,4,_,_,_,_,_]
Explanation: Your function should return k = 5, with the first five elements of nums being 0, 1, 2, 3, and 4 respectively.
It does not matter what you leave beyond the returned k (hence they are underscores).
 

Constraints:

1 <= nums.length <= 3 * 104
-100 <= nums[i] <= 100
nums is sorted in non-decreasing order.
"""

# first try
# tried to define the [-∞, i], [i, j], [j,∞] and failed for this problems
# the solution is so simple and intuitive that I never thought about
# why?
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        #[-∞, i] removed dup [i,j) dup to be handled, [unknow]
        i, j, n = 1, 1, len(nums)+1
        while i < n and j < n:
            while i < n and nums[i] != nums[i-1]:
                i += 1
            while j < n and nums[j] == nums[j-1]:
                j += 1
            nums[i] = nums[j]
        return i


# @lc code=start
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        #[-∞, i] removed dup [i,j) dup to be handled, [unknow]
        """
        i, j, n = 1, 1, len(nums)+1
        while i < n and j < n:
            while i < n and nums[i] != nums[i-1]:
                i += 1
            while j < n and nums[j] == nums[j-1]:
                j += 1
            nums[i] = nums[j]
        return i
        """
        j = 1
        for i in range(1, len(nums)):
            if nums[i] != nums[i-1]:
                nums[j] = nums[i]
                j += 1
        return j

        
# @lc code=end

